### ISSUE TITLE
Optional Column Headers for List Views

### ISSUE THREAD

**Issue by beanh66** [patternfly/patternfly-design#625]
- Enhancement on an existing pattern 
- This is currently implemented in OpenShift.io and Tectonic. 
- Depending on the content/ complexity of a list view,  column headers may be beneficial. Tectonic, OpenShift and OpenShift.io all use a similar font style for certain headings, light grey in all CAPS. We would like to contribute this optional heading style to the List View pattern.
- See attached screenshot from Tectonic:
<img width="1155" alt="screen shot 2018-03-26 at 12 26 34 pm" src="https://user-images.githubusercontent.com/18057044/37919120-09d682dc-30f1-11e8-82a2-4999a4670edf.png">
- @patternfly-design-maintainers I am working on this contribution. 

FYI @catrobson @serenamarie125 @jennyhaines

**Comment by jennyhaines** [comment]
@beanh66 - Love the style! Only concern is the color of the light grey text. It looks like you probably used pf-black-400, but pf-black-600 is the lightest that is accessible. (Which totally stinks because I like the look of the color you picked.)

**Comment by jennyhaines** [comment]
*for reference, I think you used #bbbbbb, and it should be #72767b

**Comment by mcarrano** [comment]
I have a couple of questions and things I'd like to clarify @beanh66 . First, in the screen shot you attached, I had a hard time understanding what the headers were applying to until I understood that this was basically a list nested inside another list. @jennyhaines is there something we can do visually to clarify that? 

Before we contribute this as a new pattern, I want to make sure this works as a responsive component with our existing implementation. @matthewcarleton can you take a look? Is there any reason that we can't add headers to our current list view? My only concern is that we can keep the headings aligned with the contents of the list as we resize the view. These would need to all move together

**Comment by beanh66** [comment]
@mcarrano Sorry for the confusion. To be clear the attached image in my initial post is a screenshot from the existing Tectonic console and is meant to be an example only. That is not a design we created and I am not planning on contributing a nested list, only column headers above our standard list view. Please see below for a contribution example:
![group 7](https://user-images.githubusercontent.com/18057044/37966484-a612a8b0-3196-11e8-8a1d-981c23cb0084.png)

Thanks!

**Comment by serenamarie125** [comment]
@mcarrano we will need to determine an approach for responsive behavior. IMO we should consider hiding the headers when items become stacked in rows.  We may want to consider just hiding the headers at specific breakpoints, for example at x-sm and sm.  @beanh66 do you think this would be a good idea?

**Comment by mcarrano** [comment]
@beanh66 This makes me feel better as I found the Tecktonic design very confusing. The mockup you attached above is much clearer and I agree that we should avoid nested lists, if possible. I still have questions about how this will work responsively. @serenamarie125 agree that hiding the headers when the list starts to wrap makes sense. The problem is that since the width of the list is dependent on the number of columns, it might be hard to make that work based on a system of fixed breakpoints. How did Openshift.io handle this? Can we get a look at that implementation?

@matthewcarleton would love to hear your thoughts when you get a chance. I definitely feel this is something that we should prototype before committing to a design.

**Comment by beanh66** [comment]
@matthewcarleton Is it possible to hide the headers when the list starts to wrap? If so, I'm wondering if I can move forward with this design or if we need to wait on a prototype to find out. Thanks!

**Comment by matthewcarleton** [comment]
Hey @beanh66, sorry for the late reply to this.  To answer your question, yes we can add a media query that removes the headings at a certain breakpoint.

After talking to @mcarrano @kybaker @LHinson @jennyhaines here are my concerns with the design from a development perspective. 

1. We can hide the headers but it will have to be custom for every implementation. The content length will dictate when it needs to break. There is going to be misalignment issues with the headers and their corresponding columns as the screen gets smaller.

2. The most reasonable way to build this is to use an html table. This will ensure alignment of content and be screen agnostic. The possible downfall, (depending on how you look at it) is horizontal scrolling for mobile users. I am not sure this is any worse than removing headers altogether or truncating text to the point where you can only read a small fraction of the total string.

Hopefully that helps, if you have any questions or want to talk further just let me know :)

**Comment by catrobson** [comment]
Normally we start dropping columns and content on smaller devices in product, as well as the headers.   @mindreeper2420 would you be able to demo how this works in OpenShift.io as an example?  I don't think building this as a HTML table with horizontal scroll is ideal.  The header alignment I'm not sure if that was solved in our implementation or not in OpenShift.io.

**Comment by serenamarie125** [comment]
Agreed that the HTML table won't really meet all needs.

I think that there are a couple of use cases for the List View.  For example, a compound list view would never really be able to have headers.  But there are instances of list views which necessitate this imo.  I'd love to see a demo - @mindreeper2420 maybe you could add an animated gif here?

**Comment by mindreeper2420** [comment]
@catrobson @serenamarie125 OpenShift.io currently uses the list view component from PatternFly with the simple expansion and optional headers turned on (https://rawgit.com/patternfly/patternfly-ng/master-dist/dist-demo/#/list), which also uses the responsive properties from the list view pattern. With this, the headers do disappear when on smaller screens, eliminating the need for a horizontal scroll bar.

I couldn't embed a gif, but uploaded a short clip to YouTube to demonstrate how it works: https://youtu.be/bBIFtj84jLk.

**Comment by mcarrano** [comment]
@beanh66 per today's PF Design meeting, we'd like to proceed with adding design guidance for this by adding as a variation to the List View pattern. Can you also sync up with @kybaker regarding visuals? He has a few concerns about header styles so we should do a visual design pass before finalizing. Implementation concerns will be deferred to PatternFly-Next where they can be more fully considered.

Let me know if you have questions.

**Comment by matthewcarleton** [comment]
Also just to follow what @mcarrano said, the one thing I noticed that makes this less problematic is if all content is left aligned. If the label is left aligned and the content below is right aligned it will cause alignment issues as the screen shrinks.

**Comment by beanh66** [comment]
@mcarrano sounds good, thanks for the update! I'll get with @kybaker on the visuals before posting the documentation for review.

**Comment by beanh66** [comment]
Updated the example to follow existing PF typography per Kyle's request. FYI @mceledonia @bdellasc @jennyhaines products will need to be updated from the current use of all caps for headings. 

![list-view-headers](https://user-images.githubusercontent.com/18057044/39124723-7b8e4326-46ca-11e8-8855-a3fdca9c4f99.png)
@kybaker let me know if this design works or if it needs further updates.

**Comment by jennyhaines** [comment]
@beanh66 - I believe that this treatment makes sense for the current version of PF. I do see Kyle's point around not wanting to add yet another text style if we don't have to. We will continue to have the optional column headers for PatternFly Next, but would style that text slightly differently due to our font change to Overpass, different specified weights, and base font size change.

**Comment by beanh66** [comment]
Great, thanks for the info @jennyhaines!

**Comment by serenamarie125** [comment]
@jennyhaines @beanh66 the interesting thing is that the products who are already using their own variant of this feature are all using ALL CAPS for their titles, not only for List View column headers, but for other titles & headings as well.

**Comment by LHinson** [comment]
Since there is no all caps style in the PatternFly styling definitions to date, I agree that we should proceed with the style definitions that we have today (as seen in the latest screenshot). Any decisions to introduce a new style should be done so systematically vs on a case by case basis. I would consider any considerations for a new text style out of scope for this issue.
